File: Ordina.Backend/src/Application/Orders/Ordina.Orders.Application/OrderDocumentTypes.cs
namespace Ordina.Orders.Application;

/// <summary>
/// Tipos de documento de pedido y valores asociados (reserva, presupuesto, pedido).
/// </summary>
public static class OrderDocumentTypes
{
    public const string Reservation = "Reservation";
    public const string ReservationStatus = "Reserva";
    public const string ReservationPrefix = "RES-";

    /// <summary>Legacy: solo lectura y migración.</summary>
    public const string LegacyReservation = "PendingConfirmation";
    public const string LegacyReservationStatus = "Por Confirmar";
    public const string LegacyReservationPrefix = "PCF-";

    public static bool IsReservationType(string? type) =>
        string.Equals(type, Reservation, StringComparison.OrdinalIgnoreCase)
        || string.Equals(type, LegacyReservation, StringComparison.OrdinalIgnoreCase);

    public static bool IsActiveReservationStatus(string? status) =>
        string.Equals(status, ReservationStatus, StringComparison.Ordinal)
        || string.Equals(status, LegacyReservationStatus, StringComparison.Ordinal);

    public static bool IsReservationOrderNumber(string? orderNumber)
    {
        if (string.IsNullOrWhiteSpace(orderNumber)) return false;
        var num = orderNumber.Trim().ToUpperInvariant();
        return num.StartsWith(ReservationPrefix, StringComparison.Ordinal)
            || num.StartsWith(LegacyReservationPrefix, StringComparison.Ordinal);
    }
}

File: Ordina.Frontend/src/lib/order-document-types.ts
export const ORDER_TYPE_RESERVATION = "Reservation" as const;
export const ORDER_STATUS_RESERVA = "Reserva" as const;
export const ORDER_PREFIX_RESERVATION = "RES-" as const;

/** Legacy: solo lectura y migración. */
export const LEGACY_ORDER_TYPE_RESERVATION = "PendingConfirmation" as const;
export const LEGACY_ORDER_STATUS_RESERVA = "Por Confirmar" as const;
export const LEGACY_ORDER_PREFIX_RESERVATION = "PCF-" as const;

export function isReservationType(type: string | undefined | null): boolean {
  const t = (type || "").trim();
  return (
    t === ORDER_TYPE_RESERVATION || t === LEGACY_ORDER_TYPE_RESERVATION
  );
}

export function isActiveReservationStatus(
  status: string | undefined | null,
): boolean {
  const s = (status || "").trim();
  return s === ORDER_STATUS_RESERVA || s === LEGACY_ORDER_STATUS_RESERVA;
}

export function isReservationOrderNumber(
  orderNumber: string | undefined | null,
): boolean {
  const num = (orderNumber || "").trim().toUpperCase();
  return (
    num.startsWith(ORDER_PREFIX_RESERVATION) ||
    num.startsWith(LEGACY_ORDER_PREFIX_RESERVATION)
  );
}

export function isReservationOrder(order: {
  type?: string | null;
  orderNumber?: string | null;
}): boolean {
  return isReservationType(order.type) || isReservationOrderNumber(order.orderNumber);
}

export function isActiveReservation(order: {
  type?: string | null;
  status?: string | null;
  orderNumber?: string | null;
}): boolean {
  return isReservationOrder(order) && isActiveReservationStatus(order.status);
}
